File: 03_sliding_window/05_minimum_window_substring/solution.js
/*
Sliding Window -> Minimum Window Substring
Neetcode 19 / 150
Leetcode 76

Javascript

Comments:
I need to get through sliding window.

Time: O(n)
Space: O(n)
*/

class Solution {
	/**
	 * @param {string} s
	 * @param {string} t
	 * @return {string}
	 */
	minWindow(s, t) {
		if (s.length < t.length) {
			return "";
		}

		const count = {};
		for (let char of t) {
			count[char] = (count[char] || 0) + 1;
		}

		const window = {};
		let minLen = Infinity;
		let have = 0;
		let need = Object.keys(count).length;
		let left = 0;
		let resLeft = 0;
		let resRight = 0;

		for (let right = 0; right < s.length; right++) {
			window[s[right]] = (window[s[right]] || 0) + 1;

			if (s[right] in count && window[s[right]] === count[s[right]]) {
				have++;
			}

			while (have === need) {
				if (right - left + 1 < minLen) {
					minLen = right - left + 1;
					resLeft = left;
					resRight = right;
				}
				window[s[left]]--;

				if (s[left] in count && window[s[left]] < count[s[left]]) {
					have--;
				}

				left++;
			}
		}

		return minLen === Infinity ? "" : s.slice(resLeft, resRight + 1);
	}
}

File: 03_sliding_window/05_minimum_window_substring/solution.py
"""
Sliding Window -> Minimum Window Substring
Neetcode 19 / 150
Leetcode 76

Python

Comments:
I need to get through sliding window.

Time: O(n)
Space: O(n)
"""


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""

        count = {}
        for char in t:
            count[char] = 1 + count.get(char, 0)

        window = {}
        min_len = float("inf")
        have = 0
        need = len(count)
        l = 0
        res_l = 0
        res_r = 0

        for r in range(len(s)):
            window[s[r]] = 1 + window.get(s[r], 0)
            if s[r] in count and window[s[r]] == count[s[r]]:
                have += 1
            while have == need:
                if (r - l + 1) < min_len:
                    min_len = r - l + 1
                    res_l = l
                    res_r = r
                window[s[l]] -= 1
                if s[l] in count and window[s[l]] < count[s[l]]:
                    have -= 1
                l += 1
        return s[res_l : res_r + 1] if min_len != float("inf") else ""
